<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Internal functions · Island Speciation Model</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>Island Speciation Model</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../index.html">Introduction</a></li><li><a class="toctext" href="../io/index.html">Input, Output, and Settings</a></li><li class="current"><a class="toctext" href>Internal functions</a><ul class="internal"><li><a class="toctext" href="#auxfuncts.jl-1">auxfuncts.jl</a></li></ul></li><li><a class="toctext" href="../model/index.html">Main Model</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Internal functions</a></li></ul><a class="edit-page" href="https://github.com/lleiding/islspecmod/blob/master/docs/src/aux.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Internal functions</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Internal-functions-1" href="#Internal-functions-1">Internal functions</a></h1><h2><a class="nav-anchor" id="auxfuncts.jl-1" href="#auxfuncts.jl-1">auxfuncts.jl</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.checkborderconditions-Tuple{Array{Patch,1},Int64,Int64}" href="#GeMM.checkborderconditions-Tuple{Array{Patch,1},Int64,Int64}"><code>GeMM.checkborderconditions</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">checkborderconditions!(w, x, y)</code></pre><p>check if coordinates <code>x</code> and <code>y</code> lie within world <code>w</code> and correct if not, considering defined border conditions.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L211-L215">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.createchrms-Tuple{Int64,Array{GeMM.AbstractGene,1}}" href="#GeMM.createchrms-Tuple{Int64,Array{GeMM.AbstractGene,1}}"><code>GeMM.createchrms</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>createchrms(nchrms, genes) Randomly distribute the passed genes into the given number of chromosomes. Returns an array of chromosome objects.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L477-L481">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.creategenes-Tuple{Int64,Array{GeMM.Trait,1},Dict{String,Any}}" href="#GeMM.creategenes-Tuple{Int64,Array{GeMM.Trait,1},Dict{String,Any}}"><code>GeMM.creategenes</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>creategenes(ngenes, traits, settings) Randomly create a given number of gene objects, with their base sequence and associated traits. Returns the result as an array of AbstractGenes.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L442-L446">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.createind" href="#GeMM.createind"><code>GeMM.createind</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>createind(settings, marked=false) Create an individual organism with of a new species with a random genome.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L507-L510">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.createoffspring-Tuple{Integer,GeMM.Individual,GeMM.Individual,Array{String,1}}" href="#GeMM.createoffspring-Tuple{Integer,GeMM.Individual,GeMM.Individual,Array{String,1}}"><code>GeMM.createoffspring</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>createoffspring(noffs, individual, partner, traitnames) The main reproduction function. Take two organisms and create the given number of offspring individuals. Returns an array of individuals.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L331-L335">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.createtraits-Tuple{Dict{String,Any}}" href="#GeMM.createtraits-Tuple{Dict{String,Any}}"><code>GeMM.createtraits</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>createtraits(settings) Create an array of trait objects generated from the default trait values (with a random offset).</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L397-L401">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.diversity-Tuple{Array{Patch,1}}" href="#GeMM.diversity-Tuple{Array{Patch,1}}"><code>GeMM.diversity</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">diversity(w)</code></pre><p>Calculate (average) alpha, beta and gamma diversity of the world. Returns a tuple with the three values (a,b,g). cf. Veech et al. 2002</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L139-L144">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.findisland-Tuple{Array{Patch,1}}" href="#GeMM.findisland-Tuple{Array{Patch,1}}"><code>GeMM.findisland</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">findisland(w)</code></pre><p>within world <code>w</code>, find out in which direction from the continent the island(s) lie(s).</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L185-L188">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.findposspartner-Tuple{Patch,GeMM.Individual,Array{String,1}}" href="#GeMM.findposspartner-Tuple{Patch,GeMM.Individual,Array{String,1}}"><code>GeMM.findposspartner</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>findposspartner(patch, individual, traitnames) Find a reproduction partner for the given individual in the given patch.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L306-L309">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.freespace-Tuple{Array{Patch,1}}" href="#GeMM.freespace-Tuple{Array{Patch,1}}"><code>GeMM.freespace</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>freespace(world) Calculate the average amount of free space on each patch.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L169-L172">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.gausscurve" href="#GeMM.gausscurve"><code>GeMM.gausscurve</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>gausscurve(b, c, x, a=1.0) Calculate the value of the Gauss function (&quot;bell curve&quot;) at point x; with a being the maximum height of the curve, b the position of the curve center and c the standard deviation (&quot;width&quot;).</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L122-L127">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.getmeantraitvalue-Tuple{Array{GeMM.Trait,1},Integer}" href="#GeMM.getmeantraitvalue-Tuple{Array{GeMM.Trait,1},Integer}"><code>GeMM.getmeantraitvalue</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>getmeantraitvalue(traits, traitidx) Take an array of traits and return the mean value of the indexed trait.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L23-L26">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.getstdtraitvalue-Tuple{Array{GeMM.Trait,1},Integer}" href="#GeMM.getstdtraitvalue-Tuple{Array{GeMM.Trait,1},Integer}"><code>GeMM.getstdtraitvalue</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>getstdtraitvalue(traits, traitidx) Take an array of traits and return the standard deviation of the indexed trait.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L31-L34">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.gettraitdict-Tuple{Array{GeMM.AbstractGene,1},Array{String,1}}" href="#GeMM.gettraitdict-Tuple{Array{GeMM.AbstractGene,1},Array{String,1}}"><code>GeMM.gettraitdict</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>gettraitdict(genes, traitnames) Calculate the trait dict for an array of genes.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L63-L66">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.gettraitdict-Tuple{Array{GeMM.Chromosome,1},Array{String,1}}" href="#GeMM.gettraitdict-Tuple{Array{GeMM.Chromosome,1},Array{String,1}}"><code>GeMM.gettraitdict</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>gettraitdict(chromosomes, traitnames) Convert a genome (an array of chromosomes) into a dict of traits and their values.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L39-L42">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.gettraitdict-Tuple{Array{GeMM.Trait,1},Array{String,1}}" href="#GeMM.gettraitdict-Tuple{Array{GeMM.Trait,1},Array{String,1}}"><code>GeMM.gettraitdict</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>gettraitdict(traits, traitnames) Construct a trait dict from a list of Trait objects.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L78-L81">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.identifyAdults!-Tuple{Patch}" href="#GeMM.identifyAdults!-Tuple{Patch}"><code>GeMM.identifyAdults!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>identifyAdults!(patch) Build up the <code>whoiswho</code> index of individuals and species in a patch.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L261-L264">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.iscompatible-Tuple{GeMM.Individual,GeMM.Individual,Array{String,1}}" href="#GeMM.iscompatible-Tuple{GeMM.Individual,GeMM.Individual,Array{String,1}}"><code>GeMM.iscompatible</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>iscompatible(mate, individual, traitnames) Check to see whether two individual organisms are reproductively compatible.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L274-L277">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.markthem!-Tuple{Array{GeMM.Individual,1}}" href="#GeMM.markthem!-Tuple{Array{GeMM.Individual,1}}"><code>GeMM.markthem!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>markthem!(community) Set each individual in the community (= array of individuals) as &quot;marked&quot;.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L562-L565">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.markthem!-Tuple{Array{Patch,1}}" href="#GeMM.markthem!-Tuple{Array{Patch,1}}"><code>GeMM.markthem!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>markthem!(world) Set every individual in the world as &quot;marked&quot;.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L580-L583">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.markthem!-Tuple{Patch}" href="#GeMM.markthem!-Tuple{Patch}"><code>GeMM.markthem!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>markthem!(habitat) Set each individual in the given patch as &quot;marked&quot;.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L572-L575">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.meiosis-Tuple{Array{GeMM.Chromosome,1},Bool}" href="#GeMM.meiosis-Tuple{Array{GeMM.Chromosome,1},Bool}"><code>GeMM.meiosis</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>meiosis(genome, maternal) Carry out meiosis on a genome (marked as maternal or not). Returns a haploid gamete genome. (genome =&gt; array of chromosomes)</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L3-L7">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.num2seq-Tuple{Integer}" href="#GeMM.num2seq-Tuple{Integer}"><code>GeMM.num2seq</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">num2seq(n)</code></pre><p>Convert an integer into binary and then into a DNA base sequence string.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L383-L386">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.seq2bignum-Tuple{String}" href="#GeMM.seq2bignum-Tuple{String}"><code>GeMM.seq2bignum</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">seq2bignum(sequence)</code></pre><p>Convert a DNA base sequence (a string) into binary and then into an BigInt (for larger genes). This saves memory.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L369-L373">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.seq2num-Tuple{String}" href="#GeMM.seq2num-Tuple{String}"><code>GeMM.seq2num</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">seq2num(sequence)</code></pre><p>Convert a DNA base sequence (a string) into binary and then into an integer. This saves memory.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L355-L359">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.traitsexist-Tuple{Dict{String,Float64},Dict{String,Any}}" href="#GeMM.traitsexist-Tuple{Dict{String,Float64},Dict{String,Any}}"><code>GeMM.traitsexist</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>traitsexist(traits, settings) Check a trait dict to make sure it contains the full set of traitnames required by the model (as defined in the settings).</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L91-L95">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.traitsexist-Tuple{GeMM.Individual,Dict{String,Any}}" href="#GeMM.traitsexist-Tuple{GeMM.Individual,Dict{String,Any}}"><code>GeMM.traitsexist</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>traitsexist(individual, settings) Make sure an individual organism has the full set of traits required by the model (as defined in the settings).</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L105-L109">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.varyalleles!-Tuple{Array{GeMM.AbstractGene,1},Dict{String,Any},Float64}" href="#GeMM.varyalleles!-Tuple{Array{GeMM.AbstractGene,1},Dict{String,Any},Float64}"><code>GeMM.varyalleles!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>varyalleles!(genes, settings, locivar) Mutate gene traits in the passed array of genes.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L540-L543">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeMM.varyalleles!-Tuple{Array{GeMM.Chromosome,1},Dict{String,Any},Float64}" href="#GeMM.varyalleles!-Tuple{Array{GeMM.Chromosome,1},Dict{String,Any},Float64}"><code>GeMM.varyalleles!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><p>varyalleles!(chromosomes, settings, locivar) Mutate gene traits in the passed array of chromosomes.</p></div></div><a class="source-link" target="_blank" href="https://github.com/lleiding/islspecmod/blob/75b480482e9032f85ee52e4e9982ad39dacbf8b6/src/auxfuncts.jl#L551-L554">source</a></section><footer><hr/><a class="previous" href="../io/index.html"><span class="direction">Previous</span><span class="title">Input, Output, and Settings</span></a><a class="next" href="../model/index.html"><span class="direction">Next</span><span class="title">Main Model</span></a></footer></article></body></html>
